### PR TITLE
Probe: use netlink to talk to conntrack

### PR DESCRIPTION
**Pull request by bboreham**
This is an update of #2130 with all necessary* changes upstreamed (typetypetype/conntrack#12, typetypetype/conntrack#13 and typetypetype/conntrack#14).

The objective is to reduce the CPU usage of the Scope probe, because it will no longer be running a `conntrack` binary which converts netlink into text then parsing that text. 

The commits on this branch include the original #2130, rebased, then some changes to make it easier to diff that code against the upstream, then replacing with upstream, then tidying up and ensuring CI passes.  Most of that should be squashed before merging, but I like the idea that we might preserve the authorship of @josephglanville who did much of the work.

As of the time of writing I have run some small tests which suggest it broadly works; have not done any big tests or checked performance.

I looked at https://github.com/vishvananda/netlink, but the conntrack code there is rudimentary: to get equal functionality I would basically have to copy most of the implementation from `typetypetype/conntrack`.

(*) I hope I didn't miss anything!

### CODE REVIEW

**Comment by bboreham** [comment]
CI is failing on the MacOS (Darwin) build - now we have four files depending on Linux-specific code, and I'm uncertain whether it is better to stub out all of their interfaces or introduce an insulating abstraction.

**Comment by josephglanville** [comment]
Cool to see this being picked up. 👍

File: probe/endpoint/conntrack.go [hunk 1]
```diff
@@ -1,103 +1,60 @@
...
+	bufferedFlows []conntrack.Conn          // flows coming out of activeFlows spend 1 walk cycle here
 	bufferSize    int
-	args          []string
+	natOnly       bool
 	quit          chan struct{}
 }
 
 // newConntracker creates and starts a new conntracker.
-func newConntrackFlowWalker(useConntrack bool, procRoot string, bufferSize int, args ...string) flowWalker {
+func newConntrackFlowWalker(useConntrack bool, procRoot string, bufferSize int, natOnly bool) flowWalker {
```
[the remark this answers is not in the document]
Comment (bboreham) [reply probe/endpoint/conntrack.go]: I hear you, but (a) it mirrors the existing usage and (b) there's another bool parameter at the beginning of the list, which selects a specialized version inside this function.

I think it needs a larger refactor to remove the two conntrack connections, and that would be the place to fix this.

**Comment by bboreham** [comment]
@2opremio the performance is only slightly better in my trials; I don't think we have a cluster that would really stress this code.  But after adding https://github.com/typetypetype/conntrack/pull/15 the whole probe uses 10% less CPU.

**Comment by bboreham** [comment]
Just noticed a panic on exit:

```
<probe> INFO: 2018/10/19 17:37:49.797634 === received SIGINT/SIGTERM ===
*** exiting
<probe> INFO: 2018/10/19 17:37:49.797780 Success Kubernetes reflector (pods)
<probe> INFO: 2018/10/19 17:37:49.797822 Success Kubernetes reflector (jobs)
<probe> INFO: 2018/10/19 17:37:49.797836 Success Kubernetes reflector (namespaces)
<probe> INFO: 2018/10/19 17:37:49.797796 Success Kubernetes reflector (daemonsets)
<probe> INFO: 2018/10/19 17:37:49.797852 Success Kubernetes reflector (storageclasses)
<probe> INFO: 2018/10/19 17:37:49.797901 Publish loop for scope.weave.svc.cluster.local exiting
<probe> INFO: 2018/10/19 17:37:49.797913 Success Kubernetes reflector (deployments)
<probe> INFO: 2018/10/19 17:37:49.797991 Success Kubernetes reflector (statefulsets)
<probe> INFO: 2018/10/19 17:37:49.797905 conntrack exiting
<probe> INFO: 2018/10/19 17:37:49.798032 Success Kubernetes reflector (persistentvolumeclaims)
<probe> INFO: 2018/10/19 17:37:49.797936 Success Kubernetes reflector (services)
<probe> INFO: 2018/10/19 17:37:49.798097 Success Kubernetes reflector (persistentvolumes)
<probe> INFO: 2018/10/19 17:37:49.797972 Success Kubernetes reflector (cronjobs)
<probe> INFO: 2018/10/19 17:37:49.798063 Success Kubernetes reflector (nodes)
panic: bad file descriptor
goroutine 127 [running]:
github.com/weaveworks/scope/vendor/github.com/typetypetype/conntrack.FollowSize.func2(0x4a, 0xc4211f0300)
/go/src/github.com/weaveworks/scope/vendor/github.com/typetypetype/conntrack/client.go:168 +0xba
created by github.com/weaveworks/scope/vendor/github.com/typetypetype/conntrack.FollowSize
/go/src/github.com/weaveworks/scope/vendor/github.com/typetypetype/conntrack/client.go:159 +0x106
```

**Comment by bboreham** [comment]
I have updated the conntrack package to bring in the perf improvement and a fix to the panic.

**Comment by bboreham** [comment]
@2opremio any chance you can give this one more glance over so I can merge?

**Comment by bboreham** [comment]
Now I would like to get https://github.com/typetypetype/conntrack/pull/19 merged since it doesn't pick up pre-existing NAT mappings without it.

**Pull request merged**